crates/rules/src/matrices/builder/matrix_builder.rs

```diff
@@ -0,0 +1,357 @@
+        &self,
+        factor_sources: &IndexSet<FactorSourceID>,
+    ) -> IndexSet<FactorSourceInRoleBuilderValidationStatus> {
+        self.primary_role
+            .validation_for_addition_of_factor_source_for_each(
+                FactorListKind::Override,
+                factor_sources,
+            )
+    }
+
+    pub fn validation_for_addition_of_factor_source_of_kind_to_recovery_override(
+        &self,
+        factor_source_kind: FactorSourceKind,
+    ) -> RoleBuilderMutateResult {
+        self.recovery_role
+            .validation_for_addition_of_factor_source_of_kind_to_list(
+                factor_source_kind,
+                FactorListKind::Override,
+            )
+    }
+
+    pub fn validation_for_addition_of_factor_source_to_recovery_override_for_each(
+        &self,
+        factor_sources: &IndexSet<FactorSourceID>,
+    ) -> IndexSet<FactorSourceInRoleBuilderValidationStatus> {
+        self.recovery_role
+            .validation_for_addition_of_factor_source_for_each(
+                FactorListKind::Override,
+                factor_sources,
+            )
+    }
+
+    pub fn validation_for_addition_of_factor_source_of_kind_to_confirmation_override(
+        &self,
+        factor_source_kind: FactorSourceKind,
+    ) -> RoleBuilderMutateResult {
+        self.confirmation_role
+            .validation_for_addition_of_factor_source_of_kind_to_list(
+                factor_source_kind,
+                FactorListKind::Override,
+            )
+    }
+
+    pub fn validation_for_addition_of_factor_source_to_confirmation_override_for_each(
+        &self,
+        factor_sources: &IndexSet<FactorSourceID>,
+    ) -> IndexSet<FactorSourceInRoleBuilderValidationStatus> {
+        self.confirmation_role
+            .validation_for_addition_of_factor_source_for_each(
+                FactorListKind::Override,
+                factor_sources,
+            )
+    }
+
+    pub fn validate_each_role_in_isolation(&self) -> MatrixBuilderMutateResult {
+        self.primary_role
+            .validate()
+            .into_matrix_err(RoleKind::Primary)?;
+        self.recovery_role
+            .validate()
+            .into_matrix_err(RoleKind::Recovery)?;
+        self.confirmation_role
+            .validate()
+            .into_matrix_err(RoleKind::Confirmation)?;
+        Ok(())
+    }
+
+    pub fn validate(&self) -> MatrixBuilderMutateResult {
+        self.validate_each_role_in_isolation()?;
+        self.validate_combination()?;
+        Ok(())
+    }
+
+    /// Adds the factor source to the primary role threshold list.
+    pub fn add_factor_source_to_primary_threshold(
+        &mut self,
+        factor_source_id: FactorSourceID,
+    ) -> MatrixBuilderMutateResult {
+        self.primary_role
+            .add_factor_source_to_list(factor_source_id, FactorListKind::Threshold)
+            .into_matrix_err(RoleKind::Primary)?;
+
+        self.validate_combination()
```

Comment:
This allows the  `SecurityShieldBuilder` to contain an invalid Matrix combination, right?

I think the same reasoning should apply here as for role mutation, where before mutating the role a validation is performed. Or there is a reasoning to allow invalid combination?

Reply:
hmm I dont follow? :) You need to expand on that. Which function allows for an invalid matrix combination? `add_factor_source_to_primary_threshold` specifically? Or all `add_` methods?

Why do you think that? 

The bodies of these `add_factor` on this type (`MatrixBuilder`) does this:
* calls `self.****_role.add` - which is throwing and we convert the error from RoleBuilderErr -> MatrixBuilderErr - if that operation fails we error out => not "invalid Matrix combination"
* Futhermore, we end all calls with `self.validation_combination()` asserting that the newly changed state of that specific builder did not result in the matrix being in an invalid state.

Right?

Reply:
Resolved in call: Im gonna remove `self.validate_combination()` on all mutation methods in MatrixBuilder - and we only ever call `self.validate_combination()` from `build` (ofc) and `validate` method.

furthermore:
* Gonna ponder if it is a good idea to add `reset_recovery_and_confirmation_role_states` method which hosts can call when user presses BACK in NavigationBar to go back to PrimaryRole. It might not be needed if the validation state of PrimaryRole always is in isolation though. As per comment above about deleting the `self.validate_combination()`... so maybe we dont need `reset_recovery_and_confirmation_role_states`.